From d9683392aa3bd26b389c0f2b47eba036791cfea2 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Sat, 9 Dec 2017 00:04:09 +0100
Subject: [PATCH 10/13] netlink: add netlink handler for gpause (-a)

Implement "ethtool -a <dev>" subcommand using netlink interface command
ETHTOOL_CMD_GET_PARAMS with ETH_PARAMS_IM_PAUSE mask.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 ethtool.c |  3 ++-
 netlink.c | 38 ++++++++++++++++++++++++++++++++++++++
 netlink.h |  1 +
 3 files changed, 41 insertions(+), 1 deletion(-)

diff --git a/ethtool.c b/ethtool.c
index dd9fe252ea9f..b51ff586d22f 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -4609,6 +4609,7 @@ int nl_gset(struct cmd_context *ctx) { return 0; }
 int nl_sset(struct cmd_context *ctx) { return 0; }
 int nl_gcoalesce(struct cmd_context *ctx) { return 0; }
 int nl_gring(struct cmd_context *ctx) { return 0; }
+int nl_gpause(struct cmd_context *ctx) { return 0; }
 #endif
 
 static const struct option {
@@ -4632,7 +4633,7 @@ static const struct option {
 	  "		[ wol %d[/%d] |  p|u|m|b|a|g|s|d...[/p|u|m|b|a|g|s|d...] ]\n"
 	  "		[ sopass %x:%x:%x:%x:%x:%x ]\n"
 	  "		[ msglvl %d | msglvl type on|off ... [--] ]\n" },
-	{ "-a|--show-pause", 1, do_gpause, NULL,
+	{ "-a|--show-pause", 1, do_gpause, nl_gpause,
 	  "Show pause options" },
 	{ "-A|--pause", 1, do_spause, NULL,
 	  "Set pause options",
diff --git a/netlink.c b/netlink.c
index d729ce1f1a18..8f1568d39317 100644
--- a/netlink.c
+++ b/netlink.c
@@ -1755,6 +1755,28 @@ static int show_ring(struct nl_context *nlctx, const struct nlattr *nest)
 	return 0;
 }
 
+static int show_pause(struct nl_context *nlctx, const struct nlattr *nest)
+{
+	const struct nlattr *tb[ETHA_PAUSE_MAX + 1] = {};
+	DECLARE_ATTR_TB_INFO(tb);
+	int ret;
+
+	if (!nest)
+		return -EOPNOTSUPP;
+	ret = mnl_attr_parse_nested(nest, attr_cb, &tb_info);
+	if (ret < 0)
+		return ret;
+
+	printf("Pause parameters for %s:\n", nlctx->enlhdr->ifname);
+	show_bool(tb[ETHA_PAUSE_AUTONEG], "Autonegotiate:\t");
+	show_bool(tb[ETHA_PAUSE_RX], "RX:\t\t");
+	show_bool(tb[ETHA_PAUSE_TX], "TX:\t\t");
+	/* ToDo: query negotiated pause frame usage */
+	putchar('\n');
+
+	return 0;
+}
+
 static int get_params_reply_cb(const struct nlmsghdr *nlhdr, void *data)
 {
 	const struct nlattr *tb[ETHA_PARAMS_MAX + 1] = {};
@@ -1786,6 +1808,15 @@ static int get_params_reply_cb(const struct nlmsghdr *nlhdr, void *data)
 			return MNL_CB_ERROR;
 		}
 	}
+	if (!req_mask || req_mask & ETH_PARAMS_IM_PAUSE) {
+		ret = show_pause(nlctx, tb[ETHA_PARAMS_PAUSE]);
+		if ((ret < 0) && !(req_mask & ~ETH_PARAMS_IM_PAUSE)) {
+			nlctx->exit_code = 76;
+			errno = -ret;
+			perror("Cannot get device pause settings");
+			return MNL_CB_ERROR;
+		}
+	}
 
 	return MNL_CB_OK;
 }
@@ -1803,3 +1834,10 @@ int nl_gring(struct cmd_context *ctx)
 				     get_params_reply_cb,
 				     ETH_PARAMS_IM_RING, 0);
 }
+
+int nl_gpause(struct cmd_context *ctx)
+{
+	return ethnl_std_get_request(ctx, ETHTOOL_CMD_GET_PARAMS,
+				     get_params_reply_cb,
+				     ETH_PARAMS_IM_PAUSE, 0);
+}
diff --git a/netlink.h b/netlink.h
index 8b7707c9a45b..58eba5780179 100644
--- a/netlink.h
+++ b/netlink.h
@@ -12,5 +12,6 @@ int nl_gset(struct cmd_context *ctx);
 int nl_sset(struct cmd_context *ctx);
 int nl_gcoalesce(struct cmd_context *ctx);
 int nl_gring(struct cmd_context *ctx);
+int nl_gpause(struct cmd_context *ctx);
 
 #endif /* ETHTOOL_NETLINK_H__ */
-- 
2.15.1

